### PR TITLE
Make M2 haplotype and clustered events filters smarter about germline events

### LINKED ISSUE #7391: Mutect2 genotype-germline-sites filtering discrepancy
A user found an issue when running --genotype-germline-sites with Mutect2 and found that the clustered events and haplotype filters were causing false negative results. 

This request was created from a contribution made by TA on July 26, 2021 18:24 UTC.

Link: [https://gatk.broadinstitute.org/hc/en-us/community/posts/4404184803227-Mutect2-genotype-germline-sites-filtering-discrepancy-](https://gatk.broadinstitute.org/hc/en-us/community/posts/4404184803227-Mutect2-genotype-germline-sites-filtering-discrepancy-)

\--

Hi, I have a few technical questions about changes in filtered variants when running mutect2 with -genotype-germline-sites.

I ran mutect2 on matched tumor-normal data with and without -genotype-germline-sites. Everything else about these runs was the same. 

When I compared the output vcfs I noticed differences in which variants pass all filters between the two different runs. Each run had unique variants that only passed - i.e. some variants were marked as pass when mutect2 was run with -genotype-germline-sites that failed when run with standard settings, and vice-versa. 

When I looked through these variants I noticed two different patterns of unique variants:

**Unique PASS variants to genotype-germline:** the unique variants that PASSED in genotypegermline but were rejected in standard analysis failed in the standard run because of the "strand\_bias" filter. The "strand\_bias" filter marks more variants in the standard analysis than in the genotypegermline analysis. Looking through these variants on IGV, they look like they are false positives and for some reason when you run mutect2 with --genotypegermlinesites it prevents this filter from accurately working.

**Unique PASS variants to standard:** These variants were all rejected in genotypegermline but passed in standard mutect2 failed because of haplotype or clustered\_events. I believe this is a potential problem with --genotypegermlinesites because when you include germlinesites, bona fide somatic variants that happen to be close to germline sites get filtered (when you run genotype germline you are more likely to include the germline variant in the activeregion of a somatic variant because you create an active region around the germline variant in addition to the somatic variant). It seems like if you run -genotypegermline sites you will have false negatives and miss these somatic variants because they get filtered. 

These are not an insubstantial number of variants - -genotypegermline sites returned 3910 PASS variants, and there were 123 variants that failed genotypegermline sites but passed in standard mutect2 just because they failed the haplotype or clustered\_events (likely false negatives).

Do you have any suggestions for how to get around these two issues? One way I can think of to get around the second issue is to ignore the haplotype or clustered\_event filters when running --genotypegermlinesites, but this would have the effect of introducing false positives in the variant call. Is there a way to increase the number of nearby events that trigger the haplotype/clustered \_events filters? Changing this could also restore the false negatives. I am not sure how to solve the issue in which the strand\_bias filter stops working as well when running -genotypegermlinesites. 

Thank you!<br><br><i>(created from <a href='https://broadinstitute.zendesk.com/agent/tickets/171745'>Zendesk ticket #171745</a>)<br>gz#171745</i>

### PR DESCRIPTION
**Pull request by davidbenjamin**
Closes #7391

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/filtering/FilteredHaplotypeFilter.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public double calculateErrorProbability(final VariantContext vc, final Mutect2Fi
     protected void accumulateDataForLearning(final VariantContext vc, final ErrorProbabilities errorProbabilities, final Mutect2FilteringEngine filteringEngine) {
         // we record the maximum non-sequencing artifact that is not this filter itself
```
Comment (ldgauthier) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/filtering/FilteredHaplotypeFilter.java:57]: I think this should get updated to match the code update below. "Non-sequencing, non-germline"?

**Review by ldgauthier — APPROVED** [review]
How hard would it be to put in a regression test where you do something similar to #7391 and compare with and without --genotype-germline-sites?  There's not a lot of coverage for that arg as-is.  And I'm a little concerned that while the ECNT annotation changed in the SW test, there were no filtering changes in order tests.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2IntegrationTest.java [hunk 2]
```diff
@@ -1088,7 +1089,18 @@ public void testExposureOfSmithWatermanParametersIsConsistentWithPastResults(fin
...
+        // failed on the cloud.
+        final double[] outputTlods = VariantContextTestUtils.streamVcf(output)
+                        .flatMap(vc -> vc.getAttributeAsDoubleList(GATKVCFConstants.TUMOR_LOG_10_ODDS_KEY, 0).stream())
+                .mapToDouble(x -> x).toArray();
+
+        final double[] expectedTlods = VariantContextTestUtils.streamVcf(expected)
+                .flatMap(vc -> vc.getAttributeAsDoubleList(GATKVCFConstants.TUMOR_LOG_10_ODDS_KEY, 0).stream())
+                .mapToDouble(x -> x).toArray();
+
+        Assert.assertEquals(outputTlods, expectedTlods);
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2IntegrationTest.java:1103]: The `ECNT` annotation changes in some cases.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/filtering/FilteredHaplotypeFilter.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/filtering/FilteredHaplotypeFilter.java:57]: done

**Comment by davidbenjamin** [comment]
Added a regression test with two good variants that this PR rescues from being false negatives.  Also made the sensitivity thresholds in the other integration tests higher because even without `--genotype-germline-sites` this PR saves a few calls.

**Comment by ldgauthier** [comment]
I saw your "whoops" commit -- I have actually merged commits that turn off tests.  James is trying to figure out if we can have codecov check the absolute number of tests since that should almost never go down. He claims you can run select tests from the data provider with IntelliJ, but it seems... cumbersome.

**Review by ldgauthier — APPROVED** [review]
Thanks for the new test!  The sensitivity threshold is a good find too

**Pull request merged**